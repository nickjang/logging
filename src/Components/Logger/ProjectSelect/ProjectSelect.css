.project-select {
  display: flex;
  flex-direction: column;
  min-width: 20rem;
  margin-top: 2rem;
}

.project-select h2,
.project-select select {
  width: 100%;
}

.project-select select {
  font-size: 1rem;
  padding: 0.5rem;
}

.project-select select,
.project-select input {
  font-family: var(--body-font);
}

.project-select button {
  min-width: 4.5rem;
}

.select-default {
  color: darkgray;
}

@media all and (min-width: 440px) {
  .view .view-main {
    padding-left: 12px;
  }
}