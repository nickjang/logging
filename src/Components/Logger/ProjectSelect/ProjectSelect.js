import React, { Component } from 'react';
import PropTypes from 'prop-types';
import ValidationError from '../../ValidationError/ValidationError';
import LoggingContext from '../../../Context/LoggingContext';
import './ProjectSelect.css';

class ProjectSelect extends Component {
  state = {
    newProject: {
      new: false,
      value: '',
      touched: false
    },
    loading: false,
    fetchError: {
      code: '',
      message: ''
    }
  }

  static contextType = LoggingContext;

  projectRef = React.createRef();
  newProjectRef = React.createRef();

  handleNewButtonClick = (e) => {
    e.preventDefault();
    this.setState(
      { newProject: { new: true, value: '', touched: false } },
      () => { if (this.newProjectRef) this.newProjectRef.current.focus() }
    );
  }

  handleSubmitNew = (e) => {
    e.preventDefault();
    //loading/error
    //fetch(put)
    // add to state/context
    //new: false
  }

  handleCancel = (e) => {
    e.preventDefault();
    this.setState({ newProject: { new: false, value: '', touched: false } })
  }

  updateNewProject = (project) => {
    this.setState({ newProject: { new: true, value: project, touched: true } });
  }

  validateNewProject = () => {
    const newProject = this.state.newProject.value.trim();
    if (!newProject) return 'Project name cannot be empty.';
    return;
  }

  componentDidMount() {
    if (this.projectRef) this.projectRef.current.focus();
  }

  renderSelect = () => {
    const options = this.context.projects.map(project =>
      <option key={project.id} value={project.id}>{project.name}</option>);

    return (
      <h2>
        <select
          id='project-select'
          name='project'
          ref={this.projectRef}
          aria-label='Choose a project to log in.'
          aria-required='true'
          onChange={(e) => this.props.updateProject(e.target.value)}>
          {/*Automatically selects last logged project in context, or newly created project.*/}
          {options}
        </select>
        <button type='button' onClick={(e) => this.handleNewButtonClick(e)}>New</button>
      </h2>
    );
  }

  renderNew = () => {
    return (
      <>
        <output className='form-status'>{this.state.fetchError.message || (this.state.loading && 'Adding...')}</output>
        <input
          type='text'
          id='new-project'
          name='new-project'
          ref={this.newProjectRef}
          placeholder='Enter a new project title.'
          aria-label='Enter a new project title:'
          onChange={(e) => this.updateNewProject(e.target.value)} />
        <button
          type='submit'
          form='project-form'
          onClick={(e) => this.handleSubmitNew(e)}
          disabled={this.validateNewProject()}
        >Submit
        </button>
        <button type='reset' form='project-form' onClick={(e) => this.handleCancel(e)}>Cancel</button>
        {this.state.newProject.touched
          && <ValidationError errorFor='new-project' message={this.validateNewProject()} />}
      </>
    );
  }

  render() {
    return (
      <fieldset form='project-form' name='project'>
        {!this.state.newProject.new ? this.renderSelect() : this.renderNew()}
      </fieldset>
    );
  }
}

ProjectSelect.defaultProps = {
  updateProject: () => { },
  addNewProject: () => { }
}

ProjectSelect.propTypes = {
  updateProject: PropTypes.func.isRequired,
  addNewProject: PropTypes.func.isRequired
}

export default ProjectSelect;