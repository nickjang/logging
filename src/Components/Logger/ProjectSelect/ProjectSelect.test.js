import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import ProjectSelect from './ProjectSelect';

describe(`ProjectSelect component`, () => {
  it('renders ProjectSelect by default', () => {
    const wrapper = shallow(<ProjectSelect />);
    expect(toJson(wrapper)).toMatchSnapshot()
  });
})