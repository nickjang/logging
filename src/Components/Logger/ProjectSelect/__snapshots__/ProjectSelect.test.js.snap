// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectSelect component renders ProjectSelect by default 1`] = `
<fieldset
  className="lg-card project-select"
  form="project-form"
  name="project"
>
  <output
    className="form-status "
  />
  <input
    aria-label="Enter a new project title:"
    id="new-project"
    name="new-project"
    onChange={[Function]}
    placeholder="Enter a new project title."
    type="text"
  />
  <div
    className="lg-text-right mt-1"
  >
    <button
      className="lg-btn lg-btn-success mr-1"
      disabled="Project name cannot be empty."
      form="project-form"
      onClick={[Function]}
      type="submit"
    >
      Submit
    </button>
    <button
      className="lg-btn lg-btn-reset"
      disabled={true}
      form="project-form"
      onClick={[Function]}
      type="reset"
    >
      Cancel
    </button>
  </div>
</fieldset>
`;
